content/en/docs/reference/networking/virtual-ips.md

```diff
@@ -208,6 +214,15 @@ falls back to running in iptables proxy mode.
 
 {{< figure src="/images/docs/services-ipvs-overview.svg" title="Services overview diagram for IPVS proxy" class="diagram-medium" >}}
 
+### `kernelspace` proxy mode {#proxy-mode-kernelspace}
+
+_This proxy mode is only available on Windows nodes._
+
+The kube-proxy configures packet filtering rules in the Windows _Virtual Filtering Platform_,
+an extension to Windows vSwitch. These rules process encapsulated packets within the node-level
+virtual networks, and rewrite packets so that the destination IP address (and layer 2 information)
+are correct to get the packet forwarded to an endpoint of the appropriate EndpointSlice.
```

Comment:
```suggestion
The Windows VFP is analgous to tools such as IPTables, for windows. It extends the HYper-V "Switch", which initially was built to support virtual machine networking.

Unlike the kube-proxy for iptables, however, the windows kernelspace proxy has an explicit data model for expressing loadbalancers, which can be introspected using powershell commands, which we'll show below.

The architecture of the windows kernelspace proxy is similar to that of IPVS, in many ways, in that it uses an explicit kernel API (albeit a Windows one) for creating endpoints behind virtual IPs. For detailed troubleshooting, you can use a powershell terminal to run the following commands:

hnsdiag.exe list all: This is somewhat similar to the iptables-save command, or other such commands which give you the full mapping of services and endpoints in a cluster. It shows you the Networks, Namespaces, LoadBalancers, and Endpoints for your cluster.
PS C:\Users\vagrant> hnsdiag.exe list all
Networks:
Name             ID
External         32E24B48-E6C4-4447-88C3-FAA512819C7F
Calico           4F4DD175-59E0-422D-9CA7-375578445CDE

Endpoints:
Name             ID                                   Virtual Network Name
Calico_ep        36f3bcca-ea76-431f-b6b2-550783947b9c Calico
Ethernet         ae861a28-2617-4d96-a5ff-bd152bed4601 Calico
Ethernet         9a168abe-ea6d-416e-93b0-bddb876af82a Calico
Ethernet         59509278-08e5-4951-b5e9-961165978bf4 Calico
6656053c3c1c10066c7b217ad854f1c93ff0a66e54d7f462f8b944acf054eeaf_Calico c7e0e0c0-4f9b-49fa-a7e4-4e04f6dc09d2 Calico
Ethernet         fc81297f-c907-4a69-b079-7ac9b24114a7 Calico
ca38cc4b7360e3b81719c85be0f854170b94b974019637bf3fe4e06ec8870a8d_Calico 754f52d7-a5df-4775-8f47-9dcfcacce265 Calico
Ethernet         d5ca4630-a6a9-4c70-a495-c5b368e8f0e1 Calico
79afed0dbe2c7af047bd4093d55e529fa0cb4a4f41a30af0c9e54b88d77a7df1_Calico 56775725-254d-4b61-bb0a-6e24e46a15fc Calico
Ethernet         5603e8a2-1e22-46fc-9582-a3872a6ec7db Calico
Ethernet         8554c310-ed7f-4739-a460-3c1587bd1175 Calico
Ethernet         f9b78bf5-7f54-4b72-acb8-dd0f2e2341bc Calico
Ethernet         bf31b54a-0de1-4cab-ae08-15afe68f5d72 Calico
Ethernet         1243f25b-1e52-41f7-9782-fc0a7b87308e Calico
Ethernet         31c79aa4-20d3-44a8-aa65-4348f39cec5a Calico
Ethernet         31df8f2f-3f02-4566-a8b6-401627d26266 Calico
Ethernet         7f356fed-2984-4273-b73d-f8f27862d019 Calico

Namespaces:
ID                                   | Endpoint IDs
c7fcaac9-b213-4b2e-a140-cfd79c330212 | 56775725-254D-4B61-BB0A-6E24E46A15FC
27c8f6cc-0edc-4274-8f19-331569d4711b | C7E0E0C0-4F9B-49FA-A7E4-4E04F6DC09D2
a5c60663-936c-4539-87bb-0b42d45541b7 | 754F52D7-A5DF-4775-8F47-9DCFCACCE265

LoadBalancers:
ID                                   | Virtual IPs      | Direct IP IDs
1cdf12d9-9f24-49e1-94a8-9d5b89919038 |  10.96.0.10      | 31df8f2f-3f02-4566-a8b6-401627d26266 9a168abe-ea6d-416e-93b0-bddb876af82a
a60973eb-a01e-44b9-adfc-8a1a1d0897bc |  10.97.44.40     | 8554c310-ed7f-4739-a460-3c1587bd1175
8067a768-e5b3-4009-9a81-80cce52c3367 |  10.96.0.10      | d5ca4630-a6a9-4c70-a495-c5b368e8f0e1 59509278-08e5-4951-b5e9-961165978bf4
d68aa63d-3f2f-4873-85ab-b0f1810fc649 |  10.96.0.10      | 1243f25b-1e52-41f7-9782-fc0a7b87308e 31c79aa4-20d3-44a8-aa65-4348f39cec5a
d5b8a84a-de77-4b48-ba42-ddfec8e22aaf |  10.96.0.1       | bf31b54a-0de1-4cab-ae08-15afe68f5d72
05783240-a6b1-4cc5-9d73-19dd356f09cf |  10.106.242.25   | c7e0e0c0-4f9b-49fa-a7e4-4e04f6dc09d2 754f52d7-a5df-4775-8f47-9dcfcacce265 56775725-254d-4b61-bb0a-6e24e46
a15fc
Once you see all of these objects, you can look into a single one, in detail by running the following commands:

Get-HNSNetwork: This in a typical cluster might give you two networks, an "External" Network, and a "CNI" network. The latter might be named something like "Calico".

Get-HNSEndpoint: This will give you detailed information about a single endpoint in your cluster. For example, after looking at the above hnsdiag output, you can run:

PS C:\Users\vagrant> Get-HnsEndpoint 36f3bcca-ea76-431f-b6b2-550783947b9c

ActivityId         : 55A2A83C-D4A8-425D-912C-2E6565ACC979
AdditionalParams   :
EncapOverhead      : 50
Health             : @{LastErrorCode=0; LastUpdateTime=133148120141225251}
ID                 : 36F3BCCA-EA76-431F-B6B2-550783947B9C
IPAddress          : 100.244.206.66
IsRemoteEndpoint   : True
MacAddress         : 08:00:27:8d:84:d8
Name               : Calico_ep
Policies           : {@{PA=10.20.30.11; Type=PA}}
PrefixLength       : 26
Resources          : @{AdditionalParams=; AllocationOrder=1; Allocators=System.Object[]; Health=; ID=55A2A83C-D4A8-425D-912C-2E6565ACC979; PortOperationTime=0;
                     State=1; SwitchOperationTime=0; VfpOperationTime=0; parentId=1CCA005E-61CE-4B18-AB24-E90C86543FFF}
SharedContainers   : {}
State              : 1
Type               : Overlay
Version            : 38654705669
VirtualNetwork     : 4F4DD175-59E0-422D-9CA7-375578445CDE
VirtualNetworkName : Calico
To learn more about the internals of the Kube Proxy on windows, there are blog posts such as https://argonsys.com/microsoft-cloud/library/troubleshooting-kubernetes-networking-on-windows-part-1/ and https://techcommunity.microsoft.com/t5/networking-blog/windows-networking-for-kubernetes/ba-p/339667.

Note: If you are interested in running the userspace windows proxy, a port of it lives on in https://github.com/kubernetes-sigs/kpng/tree/master/backends/windows/userspace, which maybe used in windows backends which do not run in an environment with the Virtual Filtering Platform.```

Reply:
Thanks for the suggestion. It doesn't follow our docs style guide so I'm wary of accepting it as is.